---
title: "Preparing landscape genetic data by individuals"
author: "D. G. Gannon"
date: "Jan 2021"
output:
  github_document:
    pandoc_args: --webtex
bibliography: /Users/dusty/Documents/zotero_library.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)

  require(vcfR)
  require(tidyverse)
  require(sp)
  require(raster)
  require(here)

  source(
    here("R", "general_functions.R")
  )
  
```

## Load relevant data

#### Load genetic data

Load vcf file with SNPs from presumably non-coding regions and convert genotype matrix into an allelic frequency matrix ${\bf F}$, where $f_{ik}\in \{0,0.5,1\}$ for individual $i$ and locus $k$: 1 = homozygous for the randomly selected reference allele; 0.5 = heterozygous; 0 = homozygous for the alternative allele [@bradburd2018a].

```{r}

  col_vcf <- read.vcfR(here("Data", "AQFO_snps_nc.vcf"))

# Extract genotype matrix
  col_gt <- extract.gt(col_vcf, element = "GT", as.numeric = F)
  
# compute allele load
  col_alLD <- allele_load(col_vcf)

# label rows with snp IDs
  rownames(col_alLD) <- rownames(col_gt)
  
# transpose
  col_alLD <- t(col_alLD)

```

### Create a random subset of the total snps to see if results are consistent

```{r eval=FALSE}
  
# repeat the above for coding snps
  col_vcf_cd <- read.vcfR(here("Data", "AQFO_snps_cd.vcf"))

# compute allele load
  col_alLD_cd <- allele_load(col_vcf_cd)
  
# Extract genotype matrix
  col_gt_cd <- extract.gt(col_vcf_cd, element = "GT", as.numeric = F)
  
# label rows with snp IDs
  rownames(col_alLD_cd) <- rownames(col_gt_cd)
  
#transpose
  col_alLD_cd <- t(col_alLD_cd)
  
# sanity check
  all.equal(
    rownames(col_alLD),
    rownames(col_alLD_cd)
  )

# cbind the two matrices and then select a random subset of snps
  col_alLD_all <- cbind(col_alLD, col_alLD_cd)
  
# random subset
  set.seed(35261)
  rsnps <- sample(
    1:ncol(col_alLD_all),
    size = ncol(col_alLD),
    replace = F
  )
  
  col_alLD_rsub <- col_alLD_all[,rsnps]
  
# save the random subset
  save(
    col_alLD_rsub,
    file = here("Data", "allele_loads_rand_subset.RData")
  )

```


$~$

### Load field data collected on individual plants and meadows
  
  
```{r}
 
# convert to dataframe
  col_alLD_df <- as.data.frame(col_alLD) 
    col_alLD_df <- cbind(rownames(col_alLD_df), col_alLD_df)
    names(col_alLD_df)[1] <- "Sample"

# load sample data
  samp_dat <- read.table(here("Data", "sequenced_samples.txt"), 
                              header = T, as.is = T, sep = "\t")

# subset for columns of interest
  samp_dat_sub <- samp_dat[, c("Sample", "COMPLEX", "MEADOW_ID",
                               "MEADOW_AREA", "PROP_FOREST_100",
                               "X_COORD", "Y_COORD", "PL_IN_5M", "COVER",
                               "FLW_STATUS")]
  
 # merge population variables with L matrix
  col_popgen_data <- merge(samp_dat_sub[,c("Sample", "COMPLEX", "MEADOW_ID")], 
                           col_alLD_df)
  
# sort by complex then meadow
  col_popgen_data <- col_popgen_data[order(col_popgen_data$COMPLEX, 
                                           col_popgen_data$MEADOW_ID), ]
  
```


$~$

## Compute landscape variables of interest

#### Amount of intervening forest 


**Create spatial lines object** 

The spatial lines object will include a line between all pairs of individuals.

*Very slow code and did not end up using...*

```{r}

# define CRS
  hja_crs <- CRS("+proj=utm +zone=10 +ellps=WGS84 +datum=WGS84 +units=m +no_defs")
  forest_nad <- raster(here("Data","forest.tif"))
  forest_wgs <- projectRaster(forest_nad, crs = hja_crs)

# # number of meadows
#   num_plants <- nrow(col_popgen_data)
# 
# # create list to store start-end coords
#   node_connects <- vector(mode = "list")
# 
# for(i in 1:num_plants){
#   if(i < num_plants){
#     for(j in (i+1):num_plants){
#       coords <- as.matrix(rbind(samp_dat_sub[i, c("X_COORD","Y_COORD")],
#                                 samp_dat_sub[j, c("X_COORD","Y_COORD")]))
#       nm <- paste(samp_dat_sub$Sample[i], samp_dat_sub$Sample[j],
#                 sep = "-")
#       node_connects[[nm]] <- coords
#     }
#   }
# }  
#   
# #convert to lines
#   node_connects_l <- map(node_connects, ~Line(.))
#   
# # Convert the many lines to a single lines object
# 
#   node_connects_ls <- map(1:length(node_connects_l), 
#                           ~Lines(node_connects_l[[.]], ID=names(node_connects_l)[.]))
#   
# # define CRS
#   hja_crs <- CRS("+proj=utm +zone=10 +ellps=WGS84 +datum=WGS84 +units=m +no_defs")
# 
# # create one spatial lines object
#   node_connects_spl <- SpatialLines(node_connects_ls, proj4string = hja_crs)
  

#   forest_nad <- raster(here("Data","forest.tif"))
#   forest_wgs <- projectRaster(forest_nad, crs = hja_crs)
#   
# # extract proportion of forest between the nodes
#   prop_for <- raster::extract(forest_wgs, node_connects_spl, fun=mean, na.rm=T,
#                       df=T)
#   prop_for$ID <- names(node_connects_spl)
# 
# extract proportion of forest within 100 m of a plant
  plant_coords <- as.matrix(
    samp_dat_sub[,c(which(names(samp_dat_sub)%in%c("X_COORD", "Y_COORD")))]
    )

    plant_connect <-
      raster::extract(
        forest_wgs, 
        plant_coords, 
        fun=mean, na.rm=T, 
        buffer=100, 
        df=T
      )

  samp_dat_sub$PROP_FOREST_100M_PL <- plant_connect$forest
```


#### Pairwise covariate data

**Create a pairwise list for relevant data**

```{r}
#   pw_data <- cbind(str_split(prop_for$ID, pattern = "-",
#                                        n=2, simplify = T), 
#                    prop_for)
#   
#   names(pw_data) <- c("plant_i", "plant_j", "pair_ij", "prop_forest_ij")
#   
# # merge in covariates for meadow_i
#   covs <- c("Sample", "COMPLEX", "MEADOW_ID", "MEADOW_AREA",
#             "PROP_FOREST_100", "PL_IN_5M",
#             "COVER", "X_COORD", "Y_COORD")
#   pw_data_i <- merge(pw_data, 
#                    samp_dat_sub[,colnames(samp_dat_sub) %in% covs],
#                    by.x="plant_i", by.y="Sample", sort=F)
#   
# # rename with "_i" appended to names
#   names(pw_data_i)[which(names(pw_data_i) %in% covs)] <- 
#     paste(names(pw_data_i)[which(names(pw_data_i) %in% covs)], 
#           "i", sep = "_")
#   
# # merge in covariates for meadow_j
#    pw_data_ij <- merge(pw_data_i, 
#                        samp_dat_sub[,colnames(samp_dat_sub) %in% covs],
#                   by.x="plant_j", by.y="Sample", sort=F)
# 
# # rename with "_j" appended to names   
#   names(pw_data_ij)[which(names(pw_data_ij) %in% covs)] <- 
#     paste(names(pw_data_ij)[which(names(pw_data_ij) %in% covs)], 
#           "j", sep = "_")
#   
# # reorder columns, then sort
#   pw_data_ij <- pw_data_ij[,c(2,1,3:ncol(pw_data_ij))]
#   pw_data_ij <- pw_data_ij[order(pw_data_ij$MEADOW_ID_i,
#                                  pw_data_ij$plant_i,
#                                  pw_data_ij$MEADOW_ID_j,
#                                  pw_data_ij$plant_j), ]
#   
# # sanity check
#   all.equal(rownames(col_alfreq)[1:191],
#             unique(pw_data_ij$plant_i))
#   
# # geographic distance
#   pw_data_ij <- pw_data_ij %>% 
#     mutate(
#       dist_ij = sqrt((X_COORD_i - X_COORD_j)^2 +
#         (Y_COORD_i - Y_COORD_j)^2)/1000
#     )
#     

```


$~$

**Compute the geographic distance between all pairs of nodes**

```{r}

# create coordinate matrix
  coords <- as.matrix(samp_dat_sub[,names(samp_dat_sub) %in% 
                                     c("X_COORD", "Y_COORD")])
  rownames(coords) <- samp_dat_sub$Sample

# compute distance matrix and put in km
  gdist_mat <- sqrt(gdist2(coords))/1000
  rownames(gdist_mat) <- colnames(gdist_mat)

```

$~$

#### Stack covariate matrices for covariate array, ${\bf X}$

The covariate array, $X_{(n\times n \times P)}$, is the 3-dimensional array with $P$, $n \times n$ matrix slices, where $n$ is the number of nodes (samples), that contain covariate data of $P$ covariates. For example, one slice is the geographic distance matrix between samples

```{r}

# character vector of column names
  covs_list <- c("intercept", "geo_dist", "plant_density_i",
                 "cover_i","forest_100m_i",
                 "plant_density_j", "cover_j",
                 "forest_100m_j")

# create empty array
  X <- array(dim = c(nrow(gdist_mat), ncol(gdist_mat), length(covs_list)),
             dimnames = list(rownames(gdist_mat),
                             colnames(gdist_mat),
                             covs_list))
 
# fill first couple slices (already in matrix form)
  X[,,1] <- matrix(1, nrow = dim(X)[1], ncol = dim(X)[2])
  X[,,2] <- gdist_mat
  # X[,,3] <- fillMat_sym(pw_data_ij$prop_forest_ij, rows = dim(X)[1],
  #                       cols = dim(X)[2], upper = F)
  
# # Sanity check that the matrix was filled correctly
#   rand_row <- sample(1:nrow(pw_data_ij),1)
#   row_check <- which(dimnames(X)[[1]] == pw_data_ij$plant_i[rand_row])
#   col_check <- which(dimnames(X)[[2]] == pw_data_ij$plant_j[rand_row])
#   
#   all.equal(X[row_check, col_check,3],
#             pw_data_ij$prop_forest_ij[rand_row])
#   all.equal(X[row_check, col_check,2],
#             pw_data_ij$dist_ij[rand_row])
#   
#   all.equal(samp_dat_sub$Sample,
#             dimnames(X)[[1]])
  
# create remaining slices
  
 # which are covariates pertaining to the "sending" meadow?
  rem_slices <- c("PL_IN_5M", "COVER", "PROP_FOREST_100M_PL")
  
 # Loop through these, create the matrix
 #    M_k with M_k(ij)=M_k(im) for all j and m
  for(k in 1:length(rem_slices)){
    cov_k <- which(names(samp_dat_sub) == rem_slices[k])
    temp_mat <- matrix(0, nrow = dim(X)[1], ncol = dim(X)[2])
    temp_mat <- sweep(temp_mat,1,
                      samp_dat_sub[,cov_k], FUN = "+")
    X[,,k+2] <- temp_mat
  }
  
# Loop through these, create the matrix
#    M_k with M_k(ij)=M_k(im) for all j and m
  for(k in 1:length(rem_slices)){
    cov_k <- which(names(samp_dat_sub) == rem_slices[k])
    temp_mat <- matrix(0, nrow = dim(X)[1], ncol = dim(X)[2])
    temp_mat <- sweep(temp_mat,2,
                      samp_dat_sub[,cov_k], FUN = "+")
    X[,,k+2+length(rem_slices)] <- temp_mat
  }  
  
# put things on smaller scales
  X[,,"cover_i"] <- X[,,"cover_i"]/100
  X[,,"cover_j"] <- X[,,"cover_j"]/100


# Indicator variable for if from the same meadow
  
  mead <- matrix(data = 0, nrow = nrow(col_popgen_data),
                  ncol = nrow(col_popgen_data))
  for(i in 1:nrow(mead)){
    for(j in 1:ncol(mead)){
      mead[i,j] <- as.numeric(
        col_popgen_data$MEADOW_ID[i] == col_popgen_data$MEADOW_ID[j]
      )
    }
  }
  
  X <- abind::abind(X,mead, along = 3)
  
# Indicator variable for if from the same complex
  comp <- matrix(data = 0, nrow = nrow(col_popgen_data),
                  ncol = nrow(col_popgen_data))
  for(i in 1:nrow(comp)){
    for(j in 1:ncol(comp)){
      comp[i,j] <- as.numeric(
        col_popgen_data$COMPLEX[i] == col_popgen_data$COMPLEX[j]
      )
    }
  }
  
  X <- abind::abind(X,comp,along=3)
  
# name the last slices
  dimnames(X)[[3]][c(
    dim(X)[3]-1,
    dim(X)[3]
  )] <- c("I_meadow", "I_complex")
  
```

$~$

**Save all these data as they are formatted now**

```{r}

  save(samp_dat_sub, 
       col_popgen_data, X,
       col_alLD,
       file = here("Data", "Indiv_lndscp_gen.RData"))

```


$~$

**References**








