---
title: "Estimating edge weights connecting HJA columbine genetic networks"
author: "D. G. Gannon"
date: "October 2020"
output:
  github_document:
    pandoc_args: --webtex
bibliography: /Users/dusty/Documents/zotero_library.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

  require(rstan)
  require(CholWishart)

```

### Data

We have SNP data on $n=192$ *Aquilegia formosa* individuals from XX meadows, which we consider to be "sub-populations" of the H.J. Andrews *A. formosa* population. We work with a centered, euclidean distance matrix, ${\bf D}_{(n\times n)}$, computed from a genetic loading matrix $l_{(n\times \ell)}$, where $\ell$ is the number of SNP loci sequenced. Genetic loadings are coded as $l_{ik} = 0$ if individual $i$ is homozygous for the reference allele at locus $k$, $l_{ik}=1$ if individual $i$ has a heterozygous genotype at site $k$, and $l_{ik}=2$ if individual $i$ is homozygous with two copies of the alternative allele at site $k$. The data were previously filtered to include only bi-allelic loci. 

Letting $\tilde{\bf L}$ be the column-centered loadings matrix (i.e. $\tilde l_{ik} = l_{ik} - \bar{l}_{\cdot k}$, where $\bar{l}_{\cdot k}$ is the average loading at position $k$), we compute ${\bf D}$ as

$$
{\bf D} = {\bf 1}\text{diag}({\bf G})' - 2{\bf G} + \text{diag}({\bf G}){\bf 1}',
$$

where ${\bf G} = \tilde{\bf L}^\text{T} \tilde{\bf L}$ and ${\bf 1}$ is an $n$-dimensional vector of ones.

### Model

We assume that $-{\bf D} \sim \text{Wishart}(1,\ 2{\boldsymbol \Sigma})$, where

$$
\boldsymbol{\Sigma} = ({\bf M} - \rho{\bf W})^{-1}.
$$

Above, ${\bf W}_{(n\times n)}$ is the weights matrix. It determines the degree of connectivity among nodes (individuals) in a spatial network and is the parameter of interest. The parameter ${\bf M}$ is a diagonal matrix with $m_{ii} = \sum_{j=1}^nw_{ij}$ and all off-diagonal elements equal to zero, and $\rho$ controls the amount of spatial autocorrelation among the genotypes of nearby individuals. We model the weight between individuals $i$ and $j$ as a log-linear combination of covariates and regression parameters such that

$$
w_{ij} = \exp\{{\bf x}_{ij}'\boldsymbol \beta + a_i\},
$$
where ${\bf x}_{ij}$ is a vector of covariates for individuals $i$ and $j$, $\boldsymbol \beta$ is a vector of regression parameters, and $a_i$ is an effect for the individual plant. We assume $a_1,a_2,...,a_n \sim \mathcal{N}(0,\sigma^2)$. Our covariates of interest are 

* geographic distance (km)
* proportion of forested cells that intersect the Euclidean line between plants $i$ and $j$
* percent canopy cover for plant $i$
* percent canopy cover for plant $j$
* flowering plant density around plant $i$ (individuals/25$\pi \text{m}^2$)
* flowering plant density around plant $j$ (individuals/25$\pi \text{m}^2$)

### Priors

We assume the following weakly informative priors for the regression parameters, $\rho$, and $\sigma$:

$$
{\boldsymbol \beta} \sim \mathcal{N}({\bf 0}_P,\ {\bf I}_{(P\times P)}),
$$

$$
\rho \sim \mathcal{U}(0,1),
$$

and

$$
\sigma \sim \text{half-Normal}(0,2).
$$

#### Stan Model Code

```{stan output.var="wishart_edgeweight_model", eval=FALSE}

functions{

  matrix make_W(int N, vector v, real[,,] X){
    matrix[N,N] W;
    for(i in 1:N){
      if(i == 1){W[i,i] = 0;}
      else{
        for(j in 1:(i-1)){
          W[i,i] = 0;
          W[i,j] = exp(to_row_vector(X[i,j,])*v);
          W[j,i] = W[i,j];
        }
      }
    }
    return(W);
  }

}

data{

  int<lower=1> N;          //number of individuals sampled
  int<lower=1> L;          //number of SNP loci
  int<lower=1> P;          //number of landscape variables
  
  real X[N,N,P];        //design array
  matrix[N,N] D;           //distance matrix 

}

parameters{

  vector[P] beta;              //regression parameters
  real<lower=0,upper=1> rho;   //spatial dependence
  //real a[N];                   //plant effects
  //real<lower=0> sigma;         //variation of plant effects

}

transformed parameters{

  matrix[N,N] W;
  vector[N] W_sum;
  matrix[N,N] M;
  cov_matrix[N] Sigma;
  
  W = make_W(N, beta, X);
  
  for(i in 1:N){
    W_sum[i] = sum(W[i,]);
  }
  
  M = diag_matrix(W_sum);
  
  Sigma = inverse((M - (rho*W)));

}

model{
  
//priors
  beta ~ normal(0,1);
  rho ~ uniform(0,1);
  
//likelihood
  D ~ wishart(L, Sigma);


}

```









