---
title: "Estimating edge weights connecting HJA columbine genetic networks"
author: "D. G. Gannon"
date: "February 2021"
output: pdf_document
header-includes:
  \usepackage{amsmath}
  \usepackage{bm}
  \usepackage{setspace}
  \doublespacing
bibliography: /Users/dusty/Documents/zotero_library.bib
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)

  require(rstan)
  require(CholWishart)
  require(Matrix)
  require(tidyverse)
  require(here)

#  Functions

## matrix of squared distances
  sq_dist_matrix <- function(m){
    n <- nrow(m)
    ones <- rep(1,n)
    Gmat <- m%*%t(m) 
    diag_G <- diag(Gmat)
    return(ones%*%t(diag_G) -2*Gmat + diag_G%*%t(ones))
  }
  
  
# DIC function using posterior draws
  DIC_wishart <- function(S, estims, loglik){
    with(estims, {
       -2*(dWishart(S, df=kappa, Sigma = Sigma_hat, log = T) - 
          var(loglik))
    })
  }
  
# Build contrast matrix 
  contrast_mat <- function(n){
    L <- diag(x=1, nrow = n-1, ncol = n)
    Ljp1 <- diag(x=1,nrow = n-1, ncol = n-1)
    Ljp1 <- cbind(rep(0,n-1),Ljp1)
    L-Ljp1
  }

```

## Data

We have biallelic SNP genotype data from $n=192$ *Aquilegia formosa* individuals from 25 meadows, which we consider to be "sub-populations" of the H.J. Andrews *A. formosa* population. We work with the genetic distance matrix defined by ${\bf D} = ({\bf y}_i - {\bf y}_j)^T({\bf y}_i - {\bf y}_j)$ where ${\bf y}_i,\ i=1,2,...,n$ is the $K$-vector of genotypes of individual $i$ at loci $k=1,2,...,K$ coded as $y_{ik} \in \{0,1,2\}$. $y_{ik}=0$ if individual $i$ is homozygous for the major allele at locus $k$, $y_{ik}=1$ if individual $i$ is heterozygous at locus $k$ and $y_{ik}=2$ if individual $i$ is homozygous for the minor allele at locus $k$.

### Model

@mccullagh2009 showed that distance (or dissimilarity) matrices can be modeled using a *generalized Wishart* distribution, or, equivalently, that linear contrasts on distance matrices can be modeled using a Wishart distribution. @hanks2013a illustrated the relationship between this result and the *Isolation by Resistance* model of genetic differentiation [@mcrae2006], which is a theoretical construct relating migration and gene flow to random walks along edges and among nodes of a graph, where the nodes represent populations. @mcrae2006 showed that if we imagine the graph as an electrical circuit connected by a set of resistors, then the electrical resistance among nodes is proportional to $F_{ST}$, a measure of genetic differentiation among populations [@wright1922]. 

@hanks2013a further developed a spatial regression model of genetic differentiation that utilizes the connection between graph edge weights and genetic distance and the (generalized) Wishart likelihood. We fit a similar model here, with some minor modifications.

Let $w_{ij}$ be the edge weight connecting plants $i$ and $j$, which we assume is positively related to gene flow between the spatial locations of plants $i$ and $j$. We model the edgeweights using the log-linear model

$$
w_{ij} = \exp \left\{{\bf g}_{ij}^T \bm \beta +  
\frac{1}{2}({\bf x}_i + {\bf x}_j)^T \bm \gamma + {\bf Z}_i {\bm \alpha} + {\bf Z}_j {\bm \alpha} \right\},
$$
where:

* ${\bf g}_{ij}$ is a $p_g$ vector of explanatory variables that can be measured on the edge connecting nodes $i$ and $j$ (e.g., length of the edge, whether the edge crosses some biologically relevant landscape feature, etc.).

* $\bm \beta$ is a $p_g$ vector of regression coefficients.

* ${\bf x}_i$ is a $p_x$ vector of explanatory variables measured on the node (i.e., at the location of the plant, such as canopy cover above plant $i$).

* $\bm \gamma$ is $p_x$ vector of regression coefficients.

* $\underset{\scriptscriptstyle n \times m}{\bf Z}$ is an indicator matrix indicating to which of the $m$ groups (meadows in this case) individual $i=1,2,...,n$ belongs. 

* ${\bm \alpha} \sim \mathcal{N}({\bf 0}, \tau \underset{\scriptscriptstyle m \times m}{\bf I})$, $\tau > 0$, ${\bf I}$ is the identity matrix.

Let ${\bf W}$ be the matrix of edge weights with $w_{ij}=0$ if $i=j$. A *conditional autoregressive spatial* model assumes that a given measurement, $\eta_i,\ i=1,2,...,n$, in a spatial graph (a genotype in our case) is a weighted average of the values of the nodes to which it is connected, with $w_{ij}$ the weight of the influence of node $j$ on the value of node $i$. Under this model, it can be shown that the covariance matrix defined by $\text{Var}(\bm \eta) = \bm \Sigma$ is



$$
\boldsymbol{\Sigma} = ({\bf M} - \rho{\bf W})^{-1},
$$
where ${\bf M}$ is a diagonal matrix with $m_{ii} = \sum_{j=1}^n w_{ij}$ and all off-diagonal elements equal to zero, and $\rho$ controls the amount of spatial autocorrelation in genotypes. Using results from [@mccullagh2009], we define a statistic model relating ${\bf D}$, the genetic distance matrix, to $\bm \Sigma$. Specifically,

$$
{\bf L}(-{\bf D}){\bf L}^T \sim \mathcal{W}_\kappa({\bf L}2({\bf M - \rho{\bf W}})^{-1}{\bf L}^T),
$$
where $\mathcal{W}_k({\bm \Gamma})$ denotes a Wishart distribution with degrees of freedom parameter $\kappa$ and scale matrix $\bm \Gamma$.

This model differs from that of @hanks2013a in three ways. First, we include geographic distance as an explanatory variable in the log-linear model for the weights instead of multiplying the regression equation by the inverse of distance. We felt this was more natural since it allows edge weights to go to zero at great distances if the regression parameter associated with geographic distance is negative. Second, while the theoretical connection between this model and resistance distance [@mcrae2006;@hanks2013a] is stronger in the case where $\rho =1$ (i.e., an *intrinsic autoregressive model* - ICAR model), this defines a rank-deficient precision matrix. Because generalized inverses for rank-deficient matrices are known to be numerically unstable, we chose to put a prior on $\rho$ with most prior weight towards 1. This allows the model to tend towards the ICAR model without resulting in a rank-deficient precision matrix. Finally, @hanks2013a define $\kappa=K$, which is true when all loci are neutral and mutually independent. Because we do not know if this is the case for our SNP data, we assume $\kappa$ is an unknown parameter to be estimated.


```{r}
knitr::knit_exit()
```


* **Geographic distance between plants $i$ and $j$ (km)**.

* **Average flowering plant density within 5m of plants $i$ and $j$**. We expect that high flowering plant densities should attract more pollinators. Therefore, a given plant in a high density patch may export more pollen and sire more offspring in the surrounding area than a plant in a low density patch. This should result in higher allelic covariance between this plant and a randomly selected plant on the landscape.

* **Average proportion of forested cells within a 500m radius around plants $i$ and $j$**. We use this as a measure of average "functional connectivity", hypothesizing that the more forest around a plant, the fewer pollinators will find it and therefore the less connected it is to other plants. We selected a cutoff of 500m based on previous work with hummingbirds which indicated a 50\% reduction in the probability of movement between two locations with an increase of 500m.

* **Average canopy cover over plants $i$ and $j$**. The logic for including this explanatory variable is that well-connected plants (those not surrounded by lots of forest) may still be growing under a tree. This may reduce the chances of being visited by a pollinator if pollinators flying overhead are less likely to see the plant.

* **Squared proportion of forested cells within a 500m radius**. Based on some data visualizations, it looks like there could be a peak of allelic covariance at intermediate levels surrounding forest. This does not seem unreasonable if plants and/or pollinators prefer some shade in the heat of the day or if hummingbirds prefer partially forested areas for the sake of nesting and perching.

* **Interaction between distance and canopy cover immediately over plants $i$ and $j$**. Because we think plants may be more difficult for a pollinator to find if growing beneath a tree, we might expect higher rates of mating between plants under the same tree (or trees), than two plants at the same distance but in the open. This should result in greater allelic covariance for two nearby plants under high canopy cover. However, two plants under different trees that are far apart may be very unlikely to mate. We therefore expect the effect of canopy cover to vary depending on geographic distance between plants $i$ and $j$.


### Priors

We assume the following weakly informative priors for the regression parameters and $\rho$:

$$
{\boldsymbol \beta} \sim \mathcal{N}({\bf 0}_P,\ {\bf I}_{(P\times P)}),
$$

and

$$
\rho \sim \mathcal{B}(2,2),
$$

where $\mathcal{B}(\alpha, \beta)$ is a Beta distribution with shape parameters $\alpha$ and $\beta$.

```{stan output.var="wishart_edgeweight_model_re", eval=FALSE}

functions{

// Function to create contrasts matrix L
  matrix contrasts(int N){
    matrix[N-1,N] L;
    for(i in 1:(N-1)){
      for(j in 1:N){
        if(i==j){L[i,j] = 1;}
        else if(j == (i+1)){L[i,j] = -1;}
        else{L[i,j] = 0;}
      }
    }
    return(L);
  }
  
// function to make weights matrix from regression array X and
// paramater vector v
  matrix make_W(int N, vector v, real[,,] X, vector a, matrix Z, real s){
    matrix[N,N] Wmat;
    for(i in 1:N){
      if(i == 1){Wmat[i,i] = 0;}
      else{
        for(j in 1:(i-1)){
          Wmat[i,i] = 0;
          Wmat[i,j] = exp(to_row_vector(X[i,j,])*v + 
                           (Z[i,]*a)*s + 
                           (Z[j,]*a)*s);
          Wmat[j,i] = Wmat[i,j];
        }
      }
    }
    return(Wmat);
  }

}


data{

  int<lower=1> N;          //number of individuals sampled
  int<lower=1> K;          //number of SNP loci
  int<lower=1> P;          //number of landscape and node variables
  int<lower=1> G;          // number of groups (meadows)
  
  real X[N,N,P];           //design array
  matrix[N,G] Z;           // random effect design matrix
  matrix[N,K] al_LD;       //allelic load matrix 

}

// Transform allelic load matrix into contrasts on distances
transformed data{

  matrix[N-1,N] L = contrasts(N);
  
  matrix[N-1,N-1] S = 2*(L*al_LD)*((L*al_LD)');

}


parameters{

  vector[P] beta;              //regression parameters
  real<lower=0,upper=1> rho;   //spatial dependence
  vector[G] alpha_raw;         // random meadow effects
  real<lower=0> tau;           // scale parameter for meadow effects
  real<lower=0,upper=1> kappa_std; // degrees of freedom parameter

}

transformed parameters{

  matrix[N,N] W;                 // spatial weights or conductance matrix
  vector[N] W_sum;               // vector of row sums of W
  matrix[N,N] M;                 // diagonal matrix with W_sum along diagonal
  cov_matrix[N] Sigma;           // inverse of (M-rho*W)
  real<lower=N,upper=K> kappa;   // scaled and shifted degrees of freedom parameter
  
  
  W = make_W(N, beta, X, alpha_raw, Z, tau);  // define W based on loglinear model
  
  for(i in 1:N){
    W_sum[i] = sum(W[i,]);
  }
  
  M = diag_matrix(W_sum);
  
  Sigma = inverse((M - (rho*W)));
  
  kappa = (K-N)*kappa_std + N;               // shift and scale degrees of freedom parameter

}

model{
  
//priors
  beta ~ normal(0,5);
  rho ~ beta(5,1);
  kappa_std ~ beta(5, 1);
  alpha_raw ~ normal(0,1);
  tau ~ normal(0,2);
  

//likelihood
  S ~ wishart(kappa, L*(2*Sigma)*(L'));

}

generated quantities{

  matrix[N-1,N-1] pred;    // Posterior predictive draws
  real loglik;             // loglikelihood of posterior draws given data
  
  loglik = wishart_lpdf(S | kappa, L*(2*Sigma)*(L'));
  pred = wishart_rng(kappa, L*(2*Sigma)*(L'));

}

```


```{r}
  load(here("Data", "Indiv_lndscp_gen.RData"))
```

## Format the design matrix into an $n\times n\times p$ array

```{r warning=FALSE}

  X_comb <- array(dim = c(dim(X)[c(1,2)],10))
  X_comb[,,c(1,2)] <- X[,,c(1,2)]
  
# now average and standardize them
  X_comb[,,3] <- (X[,,"plant_density_i"] +
    X[,,"plant_density_j"])/2
  
  X_comb[,,4] <- ((X[,,"cover_i"]+
                   X[,,"cover_j"])/2)

# isolation
  X_comb[,,5] <- ((X[,,"forest_500m_i"]+
            X[,,"forest_500m_j"])/2)
  
# Each variable may interact with distance
  X_comb[,,6] <- X_comb[,,2]*X_comb[,,3]
  X_comb[,,7] <- X_comb[,,2]*X_comb[,,4]
  X_comb[,,8] <- X_comb[,,2]*X_comb[,,5]
  
# create a within-complex effect
  complexes <- group_by(col_popgen_data, COMPLEX) %>%
    summarise(n=n())
  
  complex_submats <- map(1:nrow(complexes),
                         ~matrix(data = 1,
                                 nrow = complexes$n[.x],
                                 ncol = complexes$n[.x]))
  X_comb[,,9] <- as.matrix(bdiag(complex_submats))
  
# create within-meadow effect
  meadows <- group_by(col_popgen_data, MEADOW_ID) %>%
    summarise(n=n())
  
  meadow_submats <- map(1:nrow(meadows),
                        ~matrix(data = 1,
                                nrow = meadows$n[.x],
                                ncol = meadows$n[.x]))
  X_comb[,,10] <- as.matrix(bdiag(meadow_submats))
  
  
  
  dimnames(X_comb)[[3]] <- c("intercept",
                             "dist",
                             "avg_pl_density_ij", 
                             "avg_cover_ij",
                             "avg_forest500_ij",
                             "dens_by_dist",
                             "cover_by_dist",
                             "forest_by_dist",
                             "Isame_complex",
                             "Isame_meadow")
  

```

## Create meadow indexing matrix Z

```{r}

# Sanity check that things are in the right order
all.equal(rownames(col_alLD), col_popgen_data$Sample)

Z <- matrix(data = 0,
            nrow = nrow(col_alLD),
            ncol = length(unique(col_popgen_data$MEADOW_ID)))

mdws <- unique(col_popgen_data$MEADOW_ID)

for (i in 1:nrow(Z)) {
  col_id <- which(mdws == col_popgen_data$MEADOW_ID[i])
  Z[i, col_id] <- 1
}

```


## Fitting the model

```{r}
# list remainder of data inputs
 
  N <- dim(col_alLD)[1]
  num_covs <- dim(X_comb)[3]
  num_snps <- dim(col_alLD)[2]
  num_meadows <- length(mdws)

  mod_data <- list(N=N,
                   P=num_covs,
                   K=num_snps,
                   G=num_meadows,
                   X=X_comb,
                   Z=Z,
                   al_LD=col_alLD)
  
# fit the model
  mfit <- sampling(wishart_edgeweight_model_re,
                   data=mod_data,
                   iter=1000,
                   chains=2,
                   control=list(max_treedepth=12))
  
  saveRDS(mfit, file =
            here(
              "Data",
              "mfit_full_dist_interactions_wm_wc_re.rds"
            ))
```

```{r echo=FALSE}
mfit <- read_rds(
  here("Data", "mfit_full_dist_interactions_wm_wc_re.rds")
)
```


```{r}
# comparing models
  m1 <- read_rds(
    here("Data", 
         "mfit_full_dist_interactions_wm_wc_re.rds")
  )
  m2 <- read_rds(
    here(
      "Data",
      "mfit_full_dist_interactions_wm_wc.rds"
    )
  )
 # Genetic distance matrix
  L <- contrast_mat(n=N)
  genD_contrast <- 2*(L%*%col_alLD)%*%t(L%*%col_alLD)

# get posterior estimates for m1
  kappa1_post <- as.data.frame(
    rstan::extract(m1, pars="kappa")
  )
  Sigma1_post <- as.data.frame(
    rstan::extract(m1, pars="Sigma")
  )
  Sigma1_hat <- matrix(
    data = apply(Sigma1_post,2,mean),
    nrow = N,
    ncol = N
  )
  loglik1 <- as.data.frame(rstan::extract(m1, pars="loglik"))
  
# store in a list
  estims1 <- list(
    kappa=mean(kappa1_post$kappa),
    Sigma_hat = L%*%(2*Sigma1_hat)%*%t(L)
  )

# get posterior estimates for m2
  kappa2_post <- as.data.frame(
    rstan::extract(m2, pars="kappa")
  )
  Sigma2_post <- as.data.frame(
    rstan::extract(m2, pars="Sigma")
  )
  Sigma2_hat <- matrix(
    data = apply(Sigma2_post,2,mean),
    nrow = N,
    ncol = N
  )
# store in a list
  estims2 <- list(
    kappa=mean(kappa2_post$kappa),
    Sigma_hat = round(L%*%(2*Sigma2_hat)%*%t(L),6)
  )
# get draws from log posterior
  loglik2 <- as.data.frame(rstan::extract(m2, pars="loglik"))
  
  DIC_wishart(genD_contrast, estims = estims1,
              loglik = as.double(loglik1$loglik))
  DIC_wishart(genD_contrast, estims = estims2,
              loglik = as.double(loglik2$loglik))

```

## Posterior predictive checks

```{r}

# First build the contrasts matrix L
  L <- contrast_mat(
    mod_data$N
  )
# project the distance matrix down one dimension
  S_obs <- 2*(L%*%col_alLD)%*%t(L%*%col_alLD)   

# Extract posterior predictions
  D_proj_pp <- rstan::extract(mfit, pars = "pred")
  
# convert random subset of predictions to a list of matrices
  rsamps <- sample(
    1:dim(D_proj_pp[[1]])[1], 
    replace = F,
    size = 100
  )
  
  D_proj_pp_sub <- map(
    1:length(rsamps),
    ~as.matrix(D_proj_pp[[1]][.x,,])
  )
  
# convert to vectors of lower triangles
  preds <- map(
    D_proj_pp_sub,
    ~.x[lower.tri(.x, diag = T)]
  )
  
# convert to a matrix with columns as diagonal elements
#  of the wishart r.v.'s and
  preds_mat <- matrix(
    data = unlist(preds),
    nrow = length(preds),
    ncol = length(S_obs[lower.tri(S_obs, diag = T)]),
    byrow = T
  )
  
  obs_v_pred <- data.frame(
    obs = S_obs[lower.tri(S_obs, diag = T)],
    pred_mean = apply(
      preds_mat,
      MARGIN = 2,
      FUN = mean
    ),
    low = apply(
      preds_mat,
      MARGIN = 2,
      FUN = quantile,
      probs=0.025
    ),
    high = apply(
      preds_mat,
      MARGIN = 2,
      FUN = quantile,
      probs=0.975
    )
  )
  
  obs_v_pred$ID <- 1:nrow(obs_v_pred)
  
# plot the results
  ggplot(data = obs_v_pred)+
    geom_errorbar(aes(x=ID, ymin=low, ymax=high),
                  width=0.2)+
    geom_point(aes(x=ID, y=obs), colour="red",
               size=0.2)

# calculate proportion of observations outside the predictive range
  
  mean(
    (obs_v_pred$obs < obs_v_pred$low) |
    (obs_v_pred$obs > obs_v_pred$high)
  )
  
```





